tests/unittests/client/test_runner.py

```diff
@@ -118,6 +124,43 @@ def test_stop_after_max_trial_reached():
     assert client.status == status
 
 
+def test_interrupted_gather():
+    count = 10
+
+    runner = new_runner(0.01, n_workers=16)
+    runner.fct = function
+    client = runner.client
+
+    client.trials.extend([new_trial(i) for i in range(count, -1, -1)])
+
+    def bad_async(*args, **kwargs):
+        raise KeyboardInterrupt()
+
+    client.executor.async_get = bad_async
```

Comment:
It would be best to have a test where the KeyboardInterrupt occurs between trials in the gather loop, otherwise we could have buggy code that overwrites completed or broken trials to interrupted and the current test won't detect this. Like first two trials are set to completed and broken and third trial triggers a keyboardinterrupt. Only third and remaining trials should be set to interrupted.

Reply:
but that is not what this is testing.
Its only job is to execute the except part of this code.

```
        try:
            results = self.client.executor.async_get(
                self.futures, timeout=self.gather_timeout
            )
        except KeyboardInterrupt:
            self._release_all()
            raise
``

Reply:
if you want the gather loop to ignore keyboard interrupt until everything is processed that is a different feature

Reply:
I was only thinking of what was gathered before the KeyboardInterrupt occured. (during the current loop execution, not during previous gather() calls)

Reply:
if the exception occur during the async_get we dont get partial results
It is akin to having the interrupt just before the report_result() call.
There is no way to know we were close to the report_result call.

Reply:
Oh right, it's before the loop. But then all the rest of the gather method is not _protected_ and a keyboard interrupt at that point would leave them unreleased. 

I think this try-except should cover the whole run() execution instead of just the beginning of gather().

Reply:
but the try-except is not protecting anything if we want to protect the gather execution we need to delay the Keyboard interrupt.

We need something like https://github.com/Delaunay/cqueue/blob/master/msgqueue/backends/queue.py#L420
to delay the interrupt until we can process all the partial results

Reply:
this is done

Reply:
Great, thanks! :)